Sliding Window Decoding Support

jviszlai opened the pull request:
Addresses #299 and adds window decoding support. 

- Refactors much of `SequentialSinterDecoder` into a more general `WindowSinterDecoder` which allows only a subset of detectors in the segment/window to correspond to errors whose decoded results are committed.
- The existing `SequentialSinterDecoder` is preserved as a subclass of `WindowSinterDecoder` where all detectors in the window are part of the commit region.
- Adds a new `SlidingWindowSinterDecoder` that creates a sequence of overlapping windows from a `DetectorErrorModel` based on time coordinates.
- Minor bugfix of time coordinate code for readout detectors in memory experiments (previously set as `num_rounds` but this was additive to the `SHIFT_COORDS` in the repeat block)
- Adds example use to [examples/logical_error_rates.ipynb](../blob/window_decoding/examples/logical_error_rates.ipynb)

I'd appreciate thoughts on the interaction between `WindowSinterDecoder`, `SequentialSinterDecoder`, and `SlidingWindowSinterDecoder` currently. Specifically, constructing the windows for `SlidingWindowSinterDecoder` in `compile_decoder_for_dem` feels awkward when `WindowSinterDecoder` and `SequentialSinterDecoder` both construct it in `__init__`, but I couldn't think of a cleaner solution.

Also I think more general window decoding (e.g. parallel window decoding) should be easy to implement, but I'll save that for a separate PR.

perlinm commented:
Thanks for the input @beitom!

I've been thinking of dropping `Sinter` from these names anyways, so maybe we should go ahead and do that.  My primary hesitation has been the awkward fact that none of the decoders in `sinter.py` are `qldpc.decoders.Decoder`s, so I've tried to make an explicit distinction between a `Decoder` and and a `SinterDecoder`.

Thinking out loud, though, one possibility is to remedy the aforementioned awkward fact by just making every `SinterDecoder` a bona-fide `Decoder`.  Or rather, every `Compiled*Decoder` would be a `Decoder` that would implement a `.decode` method to decode a single shot, while a non-compiled decoder would implement `.decode`, but throw an error asking the user to compile with `.compile_decoder_for_dem`.  This would shorten otherwise lengthy names while mitigating potential confusion.

perlinm reviewed (commented):
@beitom sounds good.  I'll make a separate issue (#399) to clean up the API and naming of `SinterDecoder`s.  No need to bog down this PR with it.

Getting back to code review (@jviszlai): this PR looks to be in pretty good shape!  One final idea for substantive changes:

Using a detector coordinate works for the memory experiments produced by `qldpc`, but it would be valuable to make the `SlidingWindowDecoder` usable for externally generated circuits as well (such as those produced by [`tqec`](https://github.com/tqec/tqec)).  Maybe instead of passing a `time_idx` to `SlidingWindowDecoder` at initialization time, we can pass a `get_detector_time: Callable[[int], int] | None` (naming tentative) that maps a detector to a time coordinate.  If `None`, this would default to using the first detector coordinate.  Thoughts?

jviszlai commented:
Ok just finished some of the small changes. I might add another example to `logical_error_rates.ipynb`, but that might take a bit to run.

perlinm commented:
Taking a bit to run is okay.  We'll just save outputs to the notebook in the repo, and add a warning/comment for the user.

Anyways sorry I've been taking so long to get to this!  Lots of November deadlines for me 🙂.  I promise to get back to this soon (likely next week).

jviszlai commented:
Just added the new plot which doesn't take too long to run actually. I also had to do some debugging and added some edge case handling in case the number of rounds in the experiment is smaller than the window size.

And no worries at all! I've also been a bit swamped with deadlines.

perlinm commented:
@jviszlai I did a cleanup + simplify pass through the `SlidingWindowDecoder`, but I ended up with a different treatment of what happens with the "remainder" `num_time_steps % window_size` time steps at the end.  Specifically, your treatment seems to append the remainder to the last window, whereas I make another smaller window to cover the remainder.

A test script to compare implementations:
```python
from qldpc import circuits, codes, decoders
circuit = circuits.get_memory_experiment(codes.SurfaceCode(2), num_rounds=9)
decoder = decoders.SlidingWindowDecoder(4, 2)
decoder.compile_decoder_for_dem(circuit.detector_error_model())
for window in decoder.windows:
    print(window)
```
Previous implementation (inside [`SlidingWindowDecoder.compile_decoder_for_dem`](https://github.com/qLDPCOrg/qLDPC/blob/9afe0f2c9071a52b728a342d752ec29de12ab702/qldpc/decoders/sinter.py#L581)):
```python
        # create windows based on a time coordinate of the detectors in the DetectorErrorModel
        self.windows = []
        for detectors in self.detector_subsets or [range(dem.num_detectors)]:
            time_to_detectors: dict[int, list[int]] = collections.defaultdict(list)
            for detector in detectors:
                time_to_detectors[self.detector_to_time(detector)].append(detector)
            if max(time_to_detectors) < self.window_size:
                window_dets = [det for dets in time_to_detectors.values() for det in dets]
                commit_dets = window_dets
                self.windows.append((window_dets, commit_dets))
                continue
            for t in range(0, max(time_to_detectors) - self.window_size + 1, self.stride):
                window_dets = []
                commit_dets = []
                for i in range(t, t + self.window_size):
                    window_dets.extend(time_to_detectors[i])
                    if i < t + self.stride:
                        commit_dets.extend(time_to_detectors[i])
                self.windows.append((window_dets, commit_dets))
            for i in range(t + self.stride, t + self.window_size):
                self.windows[-1][1].extend(time_to_detectors[i])
            for i in range(t + self.window_size, max(time_to_detectors) + 1):
                self.windows[-1][0].extend(time_to_detectors[i])
                self.windows[-1][1].extend(time_to_detectors[i])
```
for which the test script prints
```
([0, 1, 2, 3], [0, 1])
([2, 3, 4, 5], [2, 3])
([4, 5, 6, 7, 8, 9], [4, 5, 6, 7, 8, 9])
```
New implementation:
```python
        # construct windows defined by "detection" and "commit" regions
        self.windows = []
        for detectors in self.detector_subsets or [range(dem.num_detectors)]:
            # collect detectors according to their time index
            time_to_dets: dict[int, list[int]] = collections.defaultdict(list)
            for detector in detectors:
                time_to_dets[self.detector_to_time(detector)].append(detector)

            # add one window at a time
            for start in range(0, max(time_to_dets) + 1, self.stride):
                window_time_to_dets = [time_to_dets[start + dt] for dt in range(self.window_size)]
                window = (  # defined by (detection, commit) regions
                    [det for dets in window_time_to_dets for det in dets],
                    [det for dets in window_time_to_dets[: self.stride] for det in dets],
                )
                self.windows.append(window)
```
which prints
```
([0, 1, 2, 3], [0, 1])
([2, 3, 4, 5], [2, 3])
([4, 5, 6, 7], [4, 5])
([6, 7, 8, 9], [6, 7])
([8, 9], [8, 9])
```
Which is the desired behavior?  And is it easy to understand why the remainder is treated the way it is?  If so, I'd like to explain this in the docstring.

beitom commented:
Hi All, not sure what the timeline is for wanting to be be merged but if it can wait 1-2 weeks I'm happy to also help review.

beitom commented on examples/logical_error_rates.ipynb:
```diff
@@ -739,13 +739,397 @@
+    "    max_shots: int = 10**5,\n",
+    "    max_errors: int = 100,\n",
+    "    distance_trials: int = 100,\n",
+    "    **decoding_kwargs: object,\n",
```
Could remove object annotations for kwargs

beitom commented on examples/logical_error_rates.ipynb:
```diff
@@ -739,13 +739,397 @@
+    "    tasks = []\n",
+    "    custom_decoders = {}  # each code is going to need its own decoder\n",
+    "    for code_index, code in enumerate(codes_to_simulate):\n",
+    "        distance = code.get_distance(bound=distance_trials)\n",
```
I noticed code.get_distance has return type `int | float` which causes type checking errors for other functions that expect int args, is there any reason distance will be a float?

perlinm replied:
This is because distance is undefined for trivial codes (with only one code word), in which case `get_distance` returns `np.nan`, which the type checker interprets as a float.  I'm open to other ideas for dealing with this edge case.

beitom replied:
For the notebook example it's fairly harmless, Ill take a look if this comes up anywhere in the source code and we can add a value error if distance comes in as nan.

```python
if isinstance(distance, float):
    raise ValueError(f"Code {code} has unknown distance")
```

I'll let you decide if you want to add this to the notebook or ignore the minor warnings

beitom commented on src/qldpc/decoders/sinter.py:
```diff
@@ -473,6 +498,101 @@ def decode_shots_to_error(
+        """
+        if not self.detector_to_time:
+            dem_coords = dem.get_detector_coordinates()
+            self.detector_to_time = lambda det: int(dem_coords[det][0])
```
error prone if called with with different DEMs will use the first DEM's coordinates since `self.detector_to_time` is set.

beitom replied:
A local variable could solve this cleanly

```python
if self.detector_to_time is not None:
    detector_to_time = self.detector_to_time
else:
    dem_coords = dem.get_detector_coordinates()
    detector_to_time = lambda det, coords=dem_coords: int(coords[det][0])
```

optional compatibility validation (not sure of this is overkill)

```python
  for detector_subset in self.detector_subsets or [range(dem.num_detectors)]:
      for det in detector_subset:
          try:
              time = detector_to_time(det)
              if not isinstance(time, (int, np.integer)):
                  raise ValueError(
                      f"detector_to_time({det}) returned {type(time).__name__}, expected int."
                  )
          except (KeyError, IndexError, TypeError) as e:
              raise ValueError(
                  f"detector_to_time({det}) failed: {e}. "
              ) from e
```

replace subsequent uses of `self.detector_to_time` with `detector_to_time`

beitom commented on src/qldpc/decoders/sinter.py:
```diff
@@ -316,39 +319,51 @@ def decode_shots(self, detection_event_data: npt.NDArray[np.uint8]) -> npt.NDArr
-    construct the syndrome for segment S_2, namely s_2 = s_1 + H @ e_1.  More generally, the
-    syndrome for segment S_(j+1) is s_(j+1) = s_j + H @ e_j = s_1 + H @ sum_(k=1)^j e_k.  After
-    decoding all segments, the net error sum_(j=1)^n e_j is used to predict observable flips.
+    A SequentialWindowDecoder splits a detector error model into (possibly overlapping) "windows".
```
Could be a nice place to draw a diagram in the docstring as is done in other parts of `qldpc`, here's a sample I made with claude if you guys like it

```
      Detectors: -------------------------------------------------------------->

      Window i:           [ ........... Detection Region ........... ]
                          [ Commit ]
                              |
                              +---> 1. Decode errors in Detection Region.
                                    2. Commit to errors in Commit Region.
                                    3. Update syndromes in future windows
                                       based on committed errors.
                                    4. Slide window forward.
                                              |
                                              v
      Window i+1:                   [ ........... Detection Region ........... ]
                                    [ Commit ]
                                        |
                                        v
                                       ...
```

jviszlai commented on src/qldpc/decoders/sinter.py:
Yes that sounds correct, thanks!

jviszlai commented on src/qldpc/circuits/memory.py:
I think the string formatting is used later in the comment block for `DEFAULT_IMMUNE_OP_TAG`.
